### PR TITLE
Clarify expectations of Flow Segment get_url

### PR DESCRIPTION
**Pull request by samdbmg**
# Details
Adds an ADR proposing some clarifications on how the `get_url` field on Flow Segments work:
- Makes `get_url` mandatory in Flow Segment responses
- Clarifies expectations around credentials when making segment requests
- Replaces `get_url` with a list of `get_urls`

# Pivotal Story (if relevant)
Story URL: https://www.pivotaltracker.com/story/show/187054769

# Related PRs
N/A

# Submitter PR Checks
_(tick as appropriate)_

- [X] PR completes task/fixes bug
- [X] API version has been incremented if necessary
- [ ] ADR status has been updated, and ADR implementation has been recorded
- [X] Documentation updated (README, etc.)
- [X] PR added to Pivotal story (if relevant)
- [ ] Follow-up stories added to Pivotal

# Reviewer PR Checks
_(tick as appropriate)_

- [ ] PR completes task/fixes bug
- [ ] Design makes sense, and fits with our current code base
- [ ] Code is easy to follow
- [ ] PR size is sensible
- [ ] Commit history is sensible and tidy

# Info on PRs
The checks above are guidelines. They don't all have to be ticked, but they should all have been considered.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,11 +54,12 @@ The store provides a mechanism to upload and register new Flow Segments, and an
 The media data contained within Flow Segments may be stored separately from the metadata linking them to a position on the timeline, separating the media data and metadata planes.
 For example our implementation uses a database (_e.g._ Amazon DynamoDB) to store Flow Segment metadata and an object store (_e.g._ AWS S3) to store the media data for Flow Segments.
 We refer to media data stored in the object store as 'media objects'.
-The Flow Segment has a single S3 download url which is the location of the media object that contains the stored media data for the Flow Segment.
+The Flow Segment has a list of S3 download urls which are the location of the media object(s) that contains the stored media data for the Flow Segment.
+If multiple URLs are returned in the list, they are assumed to be identical.
```
Comment (philipnbbc) [inline README.md:58]: `they` refers to the media object?
```suggestion
If multiple URLs are returned in the list, they are assumed to return identical media data.
```

File: api/TimeAddressableMediaStore.yaml [hunk 2]
```diff
@@ -921,6 +925,14 @@ paths:
                     subset of the samples within using `ts_offset` for the second segment in the sequence.
                   value:
                     $ref: examples/flow-segments-get-200-sample-offset.json
+                multiple_urls:
+                  summary: Multiple URLs
+                  description: |
+                    This example shows how the media object for a Flow Segment could be accessed using multiple URLs.
+                    Clients are free to choose any of the URLs given (which are all identical), however the server has
```
Comment (philipnbbc) [inline api/TimeAddressableMediaStore.yaml:932]: Same as before
```suggestion
                    Clients are free to choose any of the URLs given (which return identical media objects), however the server has
```

**Review by philipnbbc — APPROVED** [review]
Minor comprehension comment about what "identical" is referring to on a first read (happy for you to keep as-is). Otherwise LGTM

**Comment by johnbilt** [comment]
The basic API structure looks good - nice to see labels to identify what the end points are.  However at the risk of opening a can of worms what happens about adding or removing URLs?

**Comment by samdbmg** [comment]
> What happens about adding or removing URLs after the original segment record has been created?

On a "keep the core API simple" basis, I'd be inclined to say it should be an out-of-band extra that's neither required nor expressly forbidden. From a client's perspective, the list of `get_urls` they receive is a snapshot, only valid at that point in time, and may change if they make another request, but the semantics ("all URLs are equally valid and interchangeable") still apply.

From a server's perspective, it depends how you've implemented it. Either way you have to provide a `put_url` and an `object_id` to the client when they write content. The server might then choose to replicate that content somewhere else (another region?) and it can then generate `get_urls` covering both regions. Or it might be there's a CDN and/or an on-site cache involved, and it generates a pair of `get_urls`: one to go via the CDN, and one via the local cache. Or it might be you have A and B resilience ingests, leading to A and B Flows (of the same Source) which are defined to be bit-for-bit identical, and your particular TAMS store implementation responds to a signal that turns that into a separate resilient pair Flow C. Then the store lists URLs for both Flow A objects and Flow B objects in Flow C `get_urls` (unless one segment is missing, in which case you only get one: that's what the resilience was for!)

The other way to approach it would (potentially) be to allow registering additional duplicate Flow Segments (producing additional URLs), but the trade-off is it makes it easier to break immutability, because a client could come along and register a duplicate segment that differs from the original, and it would be difficult to enforce against that! IMO, we should leave it as is until the need arises to relax that restriction

**Pull request merged**